doc/stm/index.mld

```diff
@@ -0,0 +1,581 @@
+
+{[
+  let remove a t =
+    Mutex.lock t.mutex;
+    let r =
+      if mem_non_lock a t then begin
+        t.content <- S.remove a t.content;
+        (* t.cardinal <- t.cardinal - 1; *)
+        Some a
+      end
+      else None
+    in
+    Mutex.unlock t.mutex;
+    r
+]}
+
+most of the time, [qcheck-stm] won't be able to spot the bug:
+
+{[
+$ dune exec ./stm-doc/mutable_set_v4.exe
+random seed: 423411827
+generated error fail pass / total     time test name
+[✓]  100    0    0  100 /  100     0.8s STM sequential tests
+================================================================================
+success (ran 1 tests)
+]}
+
+The reason why is that there is very little chance to generate an argument for
+[Remove] corresponding to an element already added to the set in the same
+program. So the conditional branch where the mistake has been made is never
+explored. But be sure that it will be in real life!
+
+In order to overcome this problem, it is possible to skew the distribution of
+generated [cmd]s by looking at the [state] parameter. If the model of the set
+is empty, then we just generate a random element. But if there are already some
+elements in the set, then we can choose between picking one of them or generating a
+random new one:
+
+{[
+  let arb_cmd state =
+    let gen =
+      match state with
+      | [] -> Gen.int
+      | xs -> Gen.(oneof [oneofl xs; int])
+    in
+    QCheck.make ~print:show_cmd
+      (QCheck.Gen.oneof
+        [Gen.return Cardinal;
+         Gen.map (fun i -> Mem i) Gen.int;
+         Gen.map (fun i -> Add i) Gen.int;
+         Gen.map (fun i -> Remove i) gen;
+        ])
+]}
+
+This will be enough to trigger a failure in the test which can help us correct
+our implementation.
+
+{[
+random seed: 284906494
+generated error fail pass / total     time test name
+[✗]    2    0    1    1 /  100     0.0s STM sequential tests
+
+--- Failure --------------------------------------------------------------------
+
+Test STM sequential tests failed (11 shrink steps):
+
+[(Add 2078704297448381768); (Remove 2078704297448381768); (Mem 5022333781122650); Cardinal]
+
++++ Messages ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Messages for test STM sequential tests:
+
+  Results incompatible with model
+
+   (Add 2078704297448381768) : ()
+   (Remove 2078704297448381768) : Some (2078704297448381768)
+   (Mem 5022333781122650) : false
+   Cardinal : 1
+================================================================================
+failure (1 tests failed, 0 tests errored, ran 1 tests)
+]}
+
+We see in the output that there is indeed a [Remove] of a previously added
+element. So this execution path is
+now explored. As there is only one [Add], one [Remove] of the added element and
+then [Cardinal] returns [1], we can easily conclude there is something wrong in
+the implementation of the removal of an element when the element is indeed in
+the set.
```

Comment:
I would have expected the `Mem` call to have been shrunk away in the minimal counterexample? :thinking: 

I suggest choosing another counterexample and filing an issue about this deterministic, sequential example not shrinking to the minimal one.